tokenlists/typing.py

```diff
@@ -1,92 +1,133 @@
+            raise ValueError(f"Invalid token decimals: {v}")
+
+        return v
+
+    @validator("extensions")
+    def extensions_must_contain_simple_types(cls, d: Optional[dict]) -> Optional[dict]:
+        if not d:
+            return d
 
+        # `extensions` is `Dict[str, Union[str, int, bool, None]]`, but pydantic mutates entries
+        for val in d.values():
+            if not isinstance(val, (str, int, bool)) and val is not None:
+                raise ValueError(f"Incorrect extension field value: {val}")
 
-class Timestamp(DateTime):
-    def __init__(self, timestamp: str):
-        super().fromisoformat(timestamp)
+        return d
 
 
-@dc.dataclass(frozen=True, slots=True)
-class Tag:
+class Tag(BaseModel):
     name: str
     description: str
 
 
-# NOTE: Not frozen as we may need to dynamically modify this
-@dc.dataclass(slots=True)
-class TokenList:
+class TokenListVersion(BaseModel, Version):
+    major: int
+    minor: int
+    patch: int
+
+    @validator("*")
+    def no_negative_version_numbers(cls, v: int):
+        if v < 0:
+            raise ValueError("Invalid version number")
+
+        return v
+
+    # NOTE: These properties are just here so that we can use `Version` properly
+    @property
+    def prerelease(self):
+        return None
+
+    @property
+    def build(self):
+        return None
+
+    def __str__(self) -> str:
+        # NOTE: This custom string function is necessary because
+        #       `super(Version, self).__str__()` isn't working right
+        return f"{self.major}.{self.minor}.{self.patch}"
+
+
+class TokenList(BaseModel):
     name: str
-    timestamp: Timestamp
-    version: Version
+    timestamp: datetime
+    version: TokenListVersion
     tokens: List[TokenInfo]
     keywords: Optional[List[str]] = None
     tags: Optional[Dict[TagId, Tag]] = None
-    logoURI: Optional[URI] = None
-
-    def __iter__(self) -> Iterator[TokenInfo]:
-        return iter(self.tokens)
-
-    @classmethod
-    def from_dict(cls, data: Dict) -> "TokenList":
-        data = deepcopy(data)
-        data["version"] = Version(**data["version"])
-        data["tokens"] = [TokenInfo.from_dict(t) for t in data["tokens"]]
-        return cls(**data)  # type: ignore
-
-    def to_dict(self) -> Dict:
-        data = dc.asdict(self)
-        data["version"] = {
-            "major": self.version.major,
-            "minor": self.version.minor,
-            "patch": self.version.patch,
-        }
-        data["tokens"] = [t.to_dict() for t in self.tokens]
-        if self.keywords is None:
-            del data["keywords"]
-        if self.tags is None:
-            del data["tags"]
-        if self.logoURI is None or self.logoURI == "":
-            del data["logoURI"]
+    logoURI: Optional[AnyUrl] = None
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        all_tags = set(tag for token in self.tokens for tag in (token.tags if token.tags else []))
```

Comment:
This is hard for me to parse. How about:

```
tags_lists = [token.tags or [] for token in self.tokens]
all_tags = reduce(tags_lists, [])
```

Not much better for readability but it breaks up the nested list comprehension.

Reply:
fixed in e156e10